source/containers.tex

```diff
@@ -9200,6 +9238,665 @@
 \end{codeblock}
 \end{itemdescr}
 
+\rSec2[inplace.vector]{Class template \tcode{inplace_vector}}
+
+\rSec3[inplace.vector.overview]{Overview}
+
+\pnum
+\indexlibraryglobal{inplace_vector}%
+An \tcode{inplace_vector} is a contiguous container.
+Its capacity is fixed and
+its elements are stored within the \tcode{inplace_vector} object itself.
+
+\pnum
+An \tcode{inplace_vector} meets all of the requirements
+of a container\iref{container.reqmts},
+of a reversible container\iref{container.rev.reqmts},
+of a contiguous container, and
+of a sequence container,
+including most of the optional sequence container requirements\iref{sequence.reqmts}.
+The exceptions are the
+\tcode{push_front},
+\tcode{prepend_range},
+\tcode{pop_front}, and
+\tcode{emplace_front}
+member functions, which are not provided.
+Descriptions are provided here only
+for operations on \tcode{inplace_vector} that
+are not described in one of these tables or
+for operations where there is additional semantic information.
+
+\pnum
+For any \tcode{N},
+\tcode{inplace_vector<T, N>::iterator} and
+\tcode{inplace_vector<T, N>::const_iterator}
+meet the constexpr iterator requirements.
+
+\pnum
+%FIXME: Should this instead say something like:
+%FIXME:     "For any \tcode{N} such that \tcode{N > 0} is \tcode{true}, ..."?
+%FIXME: Same for similar expressions added in P0843R14?
+For any \tcode{N > 0},
+if \tcode{is_trivial_v<T>} is \tcode{false}, then
+no \tcode{inplace_vector<T, N>} member functions
+are usable in constant expressions.
```

Comment:
```suggestion
If \tcode{is_trivial_v<T>} is \tcode{false}, then
no member functions of \tcode{inplace_vector}
are usable in constant expressions.
```

Here and above, there's no need to say "for all N" nor "for all T", because we're already within the context of `inplace_vector`, i.e., the injected-class-name of some _particular_ specialization of `inplace_vector`, with a _particular_ `T` and `N`.

Note to anyone who cares: This used to say something to the effect of "no non-static member functions of", but LWG discussion in St Louis seems to have come down on the side of "who cares if e.g. `capacity` is constexpr-friendly, when `size` itself isn't," and so "no member functions [period]" is indeed what was polled and approved.

Reply:
I guess it's more appropriate to say
> no member functions of `inplace_vector` are <ins>**guaranteed to be**</ins> usable in constant expressions

or
> it is unspecified any member function of `inplace_vector` is usable in constant expressions

But this seemingly needs an LWG issue, and the sentence will be dropped if [P3074](https://wg21.link/p3074) lands.

Reply:
By my recollection, LEWG explicitly discussed the fact that library vendors have the freedom to add `noexcept` _but not to add `constexpr`_, and we were explciitly deciding to _forbid_ library vendors from making any of these functions constexpr-friendly in those cases.

That was certainly given as a rationale for why `iterator` is unconditionally constexpr-friendly: at least some of us were under the impression that if we had said "Under such-and-such conditions `iterator` does _not_ meet the constexpr iterator requirements," then we would be forcing library vendors to invent an iterator type that was "basically `T*` but constexpr-unfriendly," just for that purpose.

However, if (my recollection of) our impression was wrong — if library vendors actually _do_ have the freedom to make `inplace_vector` "more constexpr than strictly necessary" — well, IMHO that wouldn't be a bad outcome at all. :)

Reply:
I've left this as is - this wording (e.g. the use of `inplace_vector<T, N>`) is used throughout this section; to change it only here would be inconsistent, while changing it elsewhere requires additional wording.

Reply:
> Here and above, there's no need to say "for all N" nor "for all T",

But it _doesn't_ say for all N, it says for N > 0.

Reply:
This is being moot as the restriction itself is being removed by [P3074](https://wg21.link/p3074).